File: Java/valid-anagram.java
/**
* Given two strings s and t , write a function to determine if t is an anagram of s.
*
* Note:
* You may assume the string contains only lowercase alphabets.

* Follow up:
* What if the inputs contain unicode characters? How would you adapt your solution to such case?
* 
* Works for any ASCII characters - 128 ASCII chars are there.
* for unicode chars - 256 chars are there.
*/
class Solution {
    
    // 1. Brute force approach
    // Time Complexity: O(n2) | Space Complexity: O(1)
    public boolean isAnagramUsingBruteForce(String s, String t) {
        
        if (s.length() != t.length()){
            return false;
        }
        else if(s.isEmpty()){
            return true;
        }
        
        boolean isAnagram = false;
        
        boolean [] visited = new boolean[t.length()];
        
        for(int i=0; i<s.length(); i++){
            isAnagram = false;
            
            for(int j=0; j<t.length(); j++){
                if(s.charAt(i) == t.charAt(j) && !visited[j]){
                    isAnagram = true;
                    visited[j] = true;
                    break;
                }
            }
            
            if(!isAnagram){
                return false;
            }
        }
        
        return isAnagram;
    }
    
    // 2. Sorting approach
    // Time Complexity: O(nlogn) | Space Complexity: O(n) 
    public boolean isAnagramUsingSorting(String s, String t) {
        
		// length check
        if (s.length() != t.length()){
            return false;
        }
		else if(s.isEmpty()){ // empty string check
            return true;
        }
        
        char [] ch1 = s.toCharArray();
        char [] ch2 = t.toCharArray();
        
		// sort both arrays
        Arrays.sort(ch1);
        Arrays.sort(ch2);
		
        return Arrays.equals(ch1, ch2); // compare each character at same position or index
    }
    
    // 3. Efficient approach using hashing
    // Time: O(n) | Space: O(1) [because the table's size stays constant no matter how large n is]
    public boolean isAnagram(String s, String t) {
        
        int [] charCount = new int[128];
        
        for (int i=0; i<s.length(); i++){
            charCount[s.charAt(i)]++;
        }
        
        for (int i=0; i<t.length(); i++){
            charCount[t.charAt(i)]--;
        }
        
        for (int i=0;i<128;i++){
            if(charCount[i] != 0){
                return false;
            }
        }
        
        return true;
    }
}

File: Python/valid-anagram.py
'''https://leetcode.com/problems/valid-anagram/'''
#brute
'''We keep count of every distinct character in a hashmap for both S & T.
Then compare if the count for each character in both is equal'''
from typing import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s)!=len(t):
            return False
        S,T={},{}
        for i in range(0,len(s)):
            if s[i] in S.keys():
                S[s[i]]+=1 #updating count
            else:
                S[s[i]]=1 #recording a distinct element
        
        for i in range(0,len(t)):
            if t[i] in T.keys():
                T[t[i]]+=1
            else:
                T[t[i]]=1
        
        for i in range(0,len(t)):
            if t[i] in S.keys() and S[t[i]]==T[t[i]]:
                continue
            else:
                return False
        return True

#cocky way 1
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s)==Counter(t) #compares the count of every distinct element

#optimised in cocky way
'''We sort both the strings and compare them side by side.
Note that It is only space efficient but the one b4 is time efficient
also your interviewer may want you to write your own sort function for this'''
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        return sorted(s)==sorted(t)